cranelift/codegen/src/isa/aarch64/inst.isle

```diff
@@ -3458,14 +3458,26 @@
 
 (decl lower_bmask (Type Type ValueRegs) ValueRegs)
 
-;; For conversions that fit in a register, we can use csetm.
+;; For conversions smaller than a register, we need to mask off the high bits, and then
+;; we can recurse into the general case.
 ;;
-;; cmp   val, #0
+;; and   tmp, val, #ty_mask
+;; cmp   tmp, #0
 ;; csetm res, ne
 (rule 0
-      (lower_bmask (fits_in_64 _) (fits_in_64 _) val)
+      (lower_bmask out_ty (fits_in_16 in_ty) val)
+      (let ((mask_bits ImmLogic (imm_logic_from_u64 $I32 (ty_mask in_ty)))
+            (masked Reg (and_imm $I32 (value_regs_get val 0) mask_bits)))
+        (lower_bmask out_ty $I32 masked)))
+
+;; For conversions that exactly fit a register, we can use csetm.
+;;
+;; cmp   val, #0
+;; csetm res, ne
+(rule 1
```

Comment:
It would probably be worth making this rule priority 0, and the one that handles the `<= 16` case higher than the others; we'll test the arguments left-to-right, and `fits_in_64` will be true for the 16-bit and smaller cases.